### PR TITLE
Ingela/ssl/alt certs/gh 4877/otp 17475

### PR DESCRIPTION
**Pull request by IngelaAndin**
This PR makes ssl application work the same as OpenSSL s_client  for the provided example in #4877 
Regression tests and possible code documentation is still to be added.

### CODE REVIEW

**Comment by voltone** [comment]
I can confirm that this enables support for the kind of cross-signing that Let's Encrypt are doing, and that we might see elsewhere as various root CAs are expiring later this year (starting with DST Root CA X3). It handles scenarios where the old root CA is present but expired, or where it is absent from the CA bundle.

In my testing, clients that use a `partial_chain` callback to handle cross-signed certificates today will continue to work, and in the future they will be able to drop that option altogether on OTP versions that include this PR.

Thanks! 🎉 If we can get this into users' hands before September 30th I think it will prevent a lot of breakage that day. I will do my best to raise awareness in the community once this PR makes it into a release.

**Comment by IngelaAndin** [comment]
Added test case and some documentation. Still working on some further code enhancements before merging this to main and master. We are considering backporting it to the OTP-23 track, but that probably would need some further commits being backported to keep things simple.

File: lib/public_key/src/public_key.erl [hunk 1]
```diff
@@ -1110,18 +1110,18 @@ pkix_path_validation(TrustedCert, CertChain, Options)
 pkix_path_validation(#'OTPCertificate'{} = TrustedCert, CertChain, Options)
   when is_list(CertChain), is_list(Options) ->
     MaxPathDefault = length(CertChain),
-    {VerifyFun, Userstat0} =
+    {VerifyFun, UserState0} =
 	proplists:get_value(verify_fun, Options, ?DEFAULT_VERIFYFUN),
-    try pubkey_cert:validate_time(TrustedCert, Userstat0, VerifyFun) of
+    try pubkey_cert:validate_time(TrustedCert, UserState0, VerifyFun) of
         Userstate1 ->
```
Comment (zuiderkwast) [inline lib/public_key/src/public_key.erl:1116]: Userstate1 is not camel-case. Perhaps change this one too?

File: lib/ssl/src/ssl_certificate.erl [hunk 2]
```diff
@@ -295,13 +347,33 @@ do_certificate_chain(CertDbHandle, CertsDbRef, Chain, SerialNr, Issuer, _, ListD
 	    {ok, undefined, lists:reverse(Chain)}
     end.
 
+find_alternative_root([Cert | _], CertDbHandle, CertDbRef, InvalidatedList) ->
+    OtpCert = public_key:pkix_decode_cert(Cert, otp),
+    case find_issuer(#cert{der=Cert, otp=OtpCert}, CertDbHandle, CertDbRef, [], InvalidatedList) of
+        {error, issuer_not_found} ->
+            unknown_ca;
+        {ok, {SerialNr, IssuerId}} ->
+                case ssl_manager:lookup_trusted_cert(CertDbHandle, CertDbRef, SerialNr, IssuerId) of
```
Comment (zuiderkwast) [inline lib/ssl/src/ssl_certificate.erl:356]: Too much indentation here?

**Review by zuiderkwast — COMMENTED** [review]
Spotted minor style issues only. I didn't check what the PR actually does.

File: lib/public_key/src/public_key.erl (hunk 1 above)
Comment (IngelaAndin) [reply lib/public_key/src/public_key.erl:1116]: Well sure why not polish this too

File: lib/ssl/src/ssl_certificate.erl (hunk 2 above)
Comment (IngelaAndin) [reply lib/ssl/src/ssl_certificate.erl:356]: Strange Emacs agrees and that is what I use, will fix!

**Comment by IngelaAndin** [comment]
Updated after a code review, expect to merge tomorrow.  Patches can be expected during July. Some delays are expected due to summer vacations.

File: lib/ssl/test/ssl_cert_SUITE.erl [hunk 3]
```diff
@@ -98,8 +98,10 @@
          unsupported_sign_algo_cert_client_auth/1,
          longer_chain/0,
          longer_chain/1,
-         duplicate_chain/0,
-         duplicate_chain/1,
+         cross_singned_chain/0,
+         cross_singned_chain/1,
```
Comment (u3s) [inline lib/ssl/test/ssl_cert_SUITE.erl:102]: type-o, should be cross_signed in lines above?

**Review by u3s — APPROVED** [review]

File: lib/ssl/test/ssl_cert_SUITE.erl (hunk 3 above)
Comment (IngelaAndin) [reply lib/ssl/test/ssl_cert_SUITE.erl]: Fixed the typo.

File: lib/ssl/src/ssl_certificate.erl [hunk 4]
```diff
@@ -42,7 +42,8 @@
 	 select_extension/2,
 	 extensions_list/1,
 	 public_key_type/1,
-	 foldl_db/3
+	 foldl_db/3,
+         find_alternative_paths/5
```
[the remark this answers is not in the document]
Comment (zuiderkwast) [reply lib/ssl/src/ssl_certificate.erl]: It's spaces/tabs. `find_alternative_paths` has spaces while the old lines have tabs. Looks good in emacs but not in GitHub.

[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/src/ssl_certificate.erl]: @zuiderkwast  Interesting. I "folded" the export list and "re-tabed" it in emacs and now it looks good on github too.

**Pull request merged**